File: haskell/bob/Bob.hs

module Bob (responseFor) where

import Data.Char

yelling :: String -> Bool
yelling s = let c = filter isAlpha s in
    not (null c) && c == map toUpper c

responseFor :: String -> String
responseFor s
    | null (words s) = "Fine. Be that way!"
    | yelling s      = "Whoa, chill out!"
    | last s == '?'  = "Sure."
    | otherwise      = "Whatever."

File: clojure/bob/bob.clj
(ns bob)

(require '[clojure.string :as s])

(defn alpha? [char]
  (Character/isLetter char))

(defn has-alphas? [str]
  (some alpha? str))

(defn yelling? [str]
  (and (has-alphas? str)
       (= str (s/upper-case str))))

(defn question? [str]
  (= \? (last str)))

(defn response-for [str]
  (cond
    (s/blank? str)  "Fine. Be that way!"
    (yelling? str)  "Whoa, chill out!"
    (question? str) "Sure."
    :else           "Whatever."))
